### PR TITLE
Make limit on number of expanded fields configurable

### LINKED ISSUE #34778: Make limit on number of expanded fields configurable
I'd like to use multi_match to search over documents with 6600 fields. Is there any way [the 1024 limit](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html) can be increased?

### PR DESCRIPTION
**Pull request by cbuescher**
Currently we introduced a hard limit of 1024 to the number of fields a query can
be expanded to in #26541. Instead of using a hard limit, we should make this
configurable. This change removes the hard limit check and uses the existing
`max_clause_count` setting instead.

Closes #34778

### CODE REVIEW

**Review by cbuescher — COMMENTED** [review]
I just realize that `indices.query.bool.max_clause_count` isn't dynamically updateable, I don't know if we touched that point in our discussion that led to reusing this setting for the limit for the expansion of fields as well. Just would like to make whoever reviews this aware of this and ask for their opinion.

File: docs/reference/query-dsl/simple-query-string-query.asciidoc [hunk 1]
```diff
@@ -31,8 +31,9 @@ The `simple_query_string` top level parameters include:
 |`fields` |The fields to perform the parsed query against. Defaults to the
 `index.query.default_field` index settings, which in turn defaults to `*`. `*`
 extracts all fields in the mapping that are eligible to term queries and filters
-the metadata fields. There is a limit of no more than 1024 fields being queried
-at once.
+the metadata fields. There is a limit on the number of fields that can be queried
+at once which is defined by the `indices.query.bool.max_clause_count` <<search-settings>>
+which defaults to 1024.
```
Comment (jimczi) [inline docs/reference/query-dsl/simple-query-string-query.asciidoc]: Can you also add a note in the `multi_match` query ?

File: server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java [hunk 2]
```diff
@@ -51,13 +53,28 @@
 
 public class QueryStringIT extends ESIntegTestCase {
 
+    private static int CLUSTER_MAX_CLAUSE_COUNT;
+
+    @BeforeClass
+    public static void createRandomClusterSetting() {
+        CLUSTER_MAX_CLAUSE_COUNT = randomIntBetween(500, 1500);
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java]: Can we test with small numbers, this seems big for a simple ut.

**Review by jimczi — COMMENTED** [review]
Thanks @cbuescher . I left some comments. 

> I just realize that indices.query.bool.max_clause_count isn't dynamically updateable

I think it's fine. I am not even sure that we should document this extensively. Querying more than 1024 fields is an anti-pattern so new users don't need to be aware of this.

File: server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java (hunk 2 above)
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/query/QueryStringIT.java]: Sure, the test was running with the default of 1024 before, so I wanted to stay in that area but there's no need to. Will lower it to the hundreds if that sounds okay.

File: docs/reference/query-dsl/simple-query-string-query.asciidoc (hunk 1 above)
Comment (cbuescher) [reply docs/reference/query-dsl/simple-query-string-query.asciidoc]: np

**Review by jimczi — APPROVED** [review]
LGTM, thanks @cbuescher

**Pull request merged**